### PR TITLE
feature(astrojs/cloudflare): add support for `splitted` SSR bundles

### LINKED ISSUE #7462: SSR `split` does not work with Cloudflare
### What version of `astro` are you using?

2.7.0

### Are you using an SSR adapter? If so, which one?

Cloudflare

### What package manager are you using?

pnpm

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

The new SSR `split` function does not work for Cloudflare Adapter.
The documentation let's the user think it should work regardless the adapter, since it is in the general SSR Guide.

```
✘ [ERROR] Could not resolve "/home/projects/github-q6erqj/dist/$server_build/_worker.mjs"

 error   Could not resolve "/home/projects/github-q6erqj/dist/$server_build/_worker.mjs"
  File:
    :1626:15
  Code:
    1625 |   }).join("");
    > 1626 |   let error = new Error(`${text}${summary}`);
           |               ^
      1627 |   error.errors = errors;
      1628 |   error.warnings = warnings;
      1629 |   return error;
  Stacktrace:
Error: Build failed with 1 error:
error: Could not resolve "/home/projects/github-q6erqj/dist/$server_build/_worker.mjs"
    at failureErrorWithLog (file:///home/projects/github-q6erqj/node_modules/esbuild/lib/main.js:1626:15)
    at eval (file:///home/projects/github-q6erqj/node_modules/esbuild/lib/main.js:1038:25)
    at eval (file:///home/projects/github-q6erqj/node_modules/esbuild/lib/main.js:983:52)
    at buildResponseToResult (file:///home/projects/github-q6erqj/node_modules/esbuild/lib/main.js:1036:7)
    at eval (file:///home/projects/github-q6erqj/node_modules/esbuild/lib/main.js:1065:16)
    at responseCallbacks.<computed> (file:///home/projects/github-q6erqj/node_modules/esbuild/lib/main.js:687:9)
    at handleIncomingPacket (file:///home/projects/github-q6erqj/node_modules/esbuild/lib/main.js:742:9)
    at Socket.readFromStdout (file:///home/projects/github-q6erqj/node_modules/esbuild/lib/main.js:663:7)
    at EventEmitter.emit (https://githubq6erqj-sm1d.w-credentialless.staticblitz.com/blitz.131c0a4f.js:35:143925)
    at addChunk (https://githubq6erqj-sm1d.w-credentialless.staticblitz.com/blitz.131c0a4f.js:35:541673)
```

### What's the expected result?

- SSR `split` works for all Adapters with basic support.
- SSR `split` makes use of each Adapter, e.g. Cloudflare Directory Mode + `functions` folder
- (in the meantime, the docs should clarify this)

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-q6erqj?file=astro.config.mjs

### Participation

- [x] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by alexanderniebuhr**
## Changes

**DEMO: https://astro-ssr-split-directory.pages.dev**

- adds support for `build.split` in astrojs/cloudflare
- ~~does not work with prerenderd pages yet **(however that seems to be a bug in `build.split` itself https://github.com/withastro/astro/issues/7466)** @ematipico can you help?~~
- ~~is not yet tested with dynamic rest routes `[...props].astro`~~
- closes #7462 
- related to: https://github.com/withastro/astro/pull/7220
- related to: https://github.com/withastro/docs/pull/3504
- depends on: https://github.com/withastro/astro/pull/7490

```[tasklist]
- [x] make it work with prerendering
- [x] check dynamic rest parameters
- [x] improve fileURL remapping (maybe unnecessary with new hooks in future)
- [x] replace functions mode with check if directory + split is set
- [x] combine code to get rid of temporaray if branches.
- [x] improve filename replacement
- [ ] add overwrite protection, for custom added files in functions folder
- [ ] add warning if excludeMiddleware setting is not set correctly
- [ ] add support for middleware
- [ ] add custom support using assets & jSquash under _image path
- [ ] evaluate poiting eslint to the functions folder directly & use regex for renaming
```

## Testing

- added test suite & manually with cloudflare pages demo

## Docs

/cc @withastro/maintainers-docs for feedback!

### CODE REVIEW

File: packages/integrations/cloudflare/README.md
Comment (alexanderniebuhr) [inline packages/integrations/cloudflare/README.md:55]: this prob need feedback of @withastro/maintainers-docs!
How do you guys normally document experimental stuff in adapters?

File: packages/integrations/cloudflare/src/index.ts
Comment (alexanderniebuhr) [inline packages/integrations/cloudflare/src/index.ts:32]: I'm unsure why this has new formatting :/

File: packages/integrations/cloudflare/src/index.ts
Comment (alexanderniebuhr) [inline packages/integrations/cloudflare/src/index.ts:97]: maybe it is not necessary, since we can always remap the fileUrls here

File: packages/integrations/cloudflare/src/index.ts
Comment (alexanderniebuhr) [inline packages/integrations/cloudflare/src/index.ts:116]: moved this logic up here, to reuse for multiple if branches

File: packages/integrations/cloudflare/src/index.ts
Comment (alexanderniebuhr) [inline packages/integrations/cloudflare/src/index.ts:152]: this is basically the only new logic here, to support experimental (though not 100%) support.

File: packages/integrations/cloudflare/src/index.ts
Comment (alexanderniebuhr) [inline packages/integrations/cloudflare/src/index.ts:180]: this is just moved, and should not break logic

**Comment by alexanderniebuhr** [comment]
@ematipico first of all, sorry for pinging. But I would like to get this one out as soon as possible; I think having experimental support is in any way better than having no support! Iteration in the future can improve the support.

Not sure if you are the correct one to review, but if you can help with getting this on track would be amazing.

File: .changeset/healthy-books-study.md
Comment (alexanderniebuhr) [inline .changeset/healthy-books-study.md:1]: this prob need feedback of @withastro/maintainers-docs

**Comment by alexanderniebuhr** [comment]
not sure why windows test's are failing..

File: packages/integrations/cloudflare/src/index.ts
Comment (matthewp) [inline packages/integrations/cloudflare/src/index.ts:99]: Which bug?

File: packages/integrations/cloudflare/src/index.ts
Comment (matthewp) [inline packages/integrations/cloudflare/src/index.ts:104]: I'm guessing this is what is making Windows fail. I would recommend keeping everything as file URLs as long as possible, it prevents this sort of thing from happening.

Can you explain what this code is doing? It looks like it is adding the buildConfig.server to the outDir. Is that not already there?

File: packages/integrations/cloudflare/src/index.ts
Comment (ematipico) [reply packages/integrations/cloudflare/src/index.ts:99]: @alexanderniebuhr this is not a bug. The entry points, **in that hook**,  are correct.

I suppose you want to have `entryPoints` in the hook `astro:build:done`

File: packages/integrations/cloudflare/src/index.ts
Comment (alexanderniebuhr) [reply packages/integrations/cloudflare/src/index.ts:104]: it remaps the entryPoints to the correct folder, because the `URL` for the entryPoints do not take into account `server: new URL(`.${SERVER_BUILD_FOLDER}`, config.outDir),`

Give the file which is available at `astro:build:ssr`, the exposed entryPoint does not return the correct path AFAIK. Notice the difference: dist/**$server_build**/src/pages/_person_/entry._car_.astro.mjs

```
real file path: packages/integrations/cloudflare/test/fixtures/split/dist/$server_build/src/pages/_person_/entry._car_.astro.mjs

object returned by entryPoints: {
    route: '/[person]/[car]',
    type: 'page',
    pattern: /^\/([^/]+?)\/([^/]+?)\/?$/,
    segments: [ [Array], [Array] ],
    params: [ 'person', 'car' ],
    component: 'src/pages/[person]/[car].astro',
    generate: [Function (anonymous)],
    pathname: undefined,
    prerender: false
  } => URL {
    href: 'file:///Users/alexanderniebuhr/Developer/Projects/alexanderniebuhr_astro/packages/integrations/cloudflare/test/fixtures/split/dist/src/pages/_person_/entry._car_.astro.mjs',
    origin: 'null',
    protocol: 'file:',
    username: '',
    password: '',
    host: '',
    hostname: '',
    port: '',
    pathname: '/Users/alexanderniebuhr/Developer/Projects/alexanderniebuhr_astro/packages/integrations/cloudflare/test/fixtures/split/dist/src/pages/_person_/entry._car_.astro.mjs',
    search: '',
    searchParams: URLSearchParams {},
    hash: ''
  }
```

File: packages/integrations/cloudflare/src/index.ts
Comment (alexanderniebuhr) [reply packages/integrations/cloudflare/src/index.ts:99]: > @alexanderniebuhr this is not a bug. The entry points, **in that hook**, are correct.

I am quite confident that the paths returned in entryPoints are not correct. Please see https://github.com/withastro/astro/pull/7464#discussion_r1242273789 for example.

It's fine to have them in `astro:build:done`, it would also be fine to keep the `astro:buiod:ssr` step.

**Comment by alexanderniebuhr** [comment]
> Just so I understand, this is not toggling split on at the integration, but rather respecting the user's `split` value, right?

Yes this is totally right. It just makes current cloudflare adapter work with the following user options (while before it was just breaking):

```
adapter: cloudflare({ mode: 'directory' }),
build: {
    split: true,
},
```

File: packages/integrations/cloudflare/src/index.ts
Comment (ematipico) [reply packages/integrations/cloudflare/src/index.ts]: > I am quite confident that the paths returned in entryPoints are not correct

They **are correct** because they are the physical paths returned by Rollup, which is responsible for generating the files.

The issue is that later, during the build step, `astro` moves those files in `build.server` folder. I am working on adding the correct data in a proper hook.

**Comment by alexanderniebuhr** [comment]
resolved most of the comments and windows test are also all green now. ✅
However there are some more tasks/topics I would like to add (see https://github.com/withastro/astro/pull/7464#issue-1771787799)

But I think they don't have to be in this PR & could be added as a second/third iteration, WDYT?

File: packages/integrations/cloudflare/src/index.ts
Comment (alexanderniebuhr) [inline packages/integrations/cloudflare/src/index.ts:149]: This is definitely the easier API, instead of moving files. However I have not done perf benchmarks, but should be fast enough. If not, we can always go back to the old approach. let me know!

**Comment by matthewp** [comment]
Hey @alexanderniebuhr This is looking great! Have been talking about the `astro:build:ssr` having incorrect paths with @ematipico and we think we need to fix that issue first so that adapters don't need to deal with it.

Leaving a note here on that issue just because I don't think there's a better place. @ematipico in my investigation what I see is:

1. In astro:build:ssr these paths are wrong but the real files do exist inside of `build.server`.
  - This makes me think that it's a matter of the wrong paths being saved in internals.
2. Also of note, the real files are hashed filenames that are later renamed to the "pretty names" here: https://github.com/withastro/astro/blob/0daaa9efba6c7caeffcd23ce279f4f54080e9bd6/packages/astro/src/core/build/static-build.ts#L299-L301
  - Should they be renamed to the pretty names or should they rename the hash names?
  - If we need/want the prettier names, then we need to make it so that the `astro:build:ssr` hook runs after this code. Currently it does not.

File: packages/integrations/cloudflare/src/index.ts
Comment (alexanderniebuhr) [reply packages/integrations/cloudflare/src/index.ts]: _note to self: should be fixed by https://github.com/withastro/astro/pull/7483_

**Comment by alexanderniebuhr** [comment]
- I rebased on latest master, which fixes the hook path issue.
- I pushed a change which allow the usage of `[...dynamic].astro` routes
- I still a facing issue with pre-rendered pages (https://github.com/withastro/astro/issues/7466)
- I still have two DX improvements in the tasks at the top of this PR, but I would like to add them later with another PR.
- Still waiting for docs team review. 
- Would be great if we can land this as experimental, so I can try it with my personal side and find more edge cases for the second iteration.

cc @matthewp @ematipico @withastro/maintainers-docs

**Comment by ematipico** [comment]
!preview cloudflare-split

File: packages/integrations/cloudflare/src/index.ts
Comment (ematipico) [inline packages/integrations/cloudflare/src/index.ts:137]: This is a very long expression, when I read it I can't understand what is doing. Would it be possible to add some comments explaining what it's doing?

File: packages/integrations/cloudflare/src/index.ts
Comment (ematipico) [inline packages/integrations/cloudflare/src/index.ts:135]: Tip: you're creating a new array at every loop iteration, you can move `[..._entryPoints.entries()]` outside of the loop.

File: packages/integrations/cloudflare/src/index.ts
Comment (ematipico) [inline packages/integrations/cloudflare/src/index.ts:134]: Maybe we can add a comment before this loop, and explain to the fellow developers that it's doing.

File: packages/integrations/cloudflare/test/directory-split.test.js
Comment (ematipico) [inline packages/integrations/cloudflare/test/directory-split.test.js:6]:
```suggestion
describe('Cloudflare SSR split', () => {
```

File: packages/integrations/cloudflare/test/directory-split.test.js
Comment (ematipico) [inline packages/integrations/cloudflare/test/directory-split.test.js:30]:
```suggestion
	it('generates functions folders for each page emitted by astro inside the project root', async () => {
```

**Review by ematipico — APPROVED** [review]
The first round of code review. It looks good to me, nothing major from my point of view, and the tests seem solid. 

Let's wait for the docs team

**Comment by ematipico** [comment]
By the way, @alexanderniebuhr , this contribution of yours is fantastic! Great work, and thank you! 🚀

File: .changeset/healthy-books-study.md
Comment (alexanderniebuhr) [reply .changeset/healthy-books-study.md:5]: because it does not include all the tasks from the [tasklist](https://github.com/withastro/astro/pull/7464#issue-1771787799), and I want to improve the implementation in the future :)

Comment (ematipico) [reply .changeset/healthy-books-study.md:5]: I am not very familiar with experimental flags inside adapters, but if we were to add one, we should pass an explicit property, like:

```js
// astro.config.mjs
import cloudflare from "@astrojs/cloudflare"
export default defineConfig({
	output: "server",
	adapter: cloudflare({
		mode: 'directory',
		experimentalSplit: true
	})
})
```

This is the pattern that we have in core, so we could extend it in adapters too.

File: packages/integrations/cloudflare/README.md
Comment (sarah11918) [inline packages/integrations/cloudflare/README.md:51]:
```suggestion
In directory mode, the adapter will compile the client side part of your app the same way by default, but moves the worker script into a `functions` folder in the project root. In this case, the adapter will only ever place a `[[path]].js` in that folder, allowing you to add additional plugins and pages middleware which can be checked into version control.
```
P. S. We don't refer to things as new, because that "gets old" quickly!  :smile: 

And, *especially* if this is intended to be experimental, but even if not, I would tend to instead just describe the above behaviour as "default" (since by default, `split` will not be true). The paragraph below can describe what happens if someone has actively configured `split: true`.

File: packages/integrations/cloudflare/README.md
Comment (sarah11918) [inline packages/integrations/cloudflare/README.md:53]:
```suggestion
With the build configuration `split: true`, the adapter instead compiles a seperate bundle for each page which is used by cloudflare routing. Cloudflare documentation contains more information about [writing custom functions](https://developers.cloudflare.com/pages/platform/functions/).
```
I think this might be clear enough without needing to overload the first paragraph with information about `build.split`, since people might not even set that configuration?

If this is determined to be experimental (not just "new, and I hope it doesn't break" but rather behind a flag), then ping Docs again to determine the appropriate syntax for marking as experimental!

**Review by sarah11918 — COMMENTED** [review]
Thanks for this helpful contribution, @alexanderniebuhr! I'm sure people are going to appreciate this! 

Left some suggestions for your consideration based on:

- Is this actually "experimental" (i.e. behind a flag?) vs just "new"?
- Changesets are now being used to autogenerate blog post material, so congrats, this PR is suffering through those growing pains of changeset editing, not just README editing!  :smile:
- suggestion to maybe not refer to the optional config so early on, front-loading an already kind of big/explainy paragraph of the default behavior, and letting the 2nd paragraph do the heavy lifting?

File: .changeset/healthy-books-study.md
Comment (alexanderniebuhr) [reply .changeset/healthy-books-study.md:5]: ok maybe we should not say it is `experimental`, because it actually does not have a current working solution.

_(changeset discussion below)_

Comment (alexanderniebuhr) [reply .changeset/healthy-books-study.md:5]: However if you want to keep it simpler, I will use your suggestion.

Comment (ematipico) [reply .changeset/healthy-books-study.md:5]: We can add all these warnings in the `README.md`; the changeset can link to that paragraph. The changelogs get lost, but the `README.md` sticks, even on the npm page of the package.

File: packages/integrations/cloudflare/README.md
Comment (alexanderniebuhr) [inline packages/integrations/cloudflare/README.md:55]: Added some more information, so users know what to expect, regarding features.
can you check this, again? @sarah11918

**Review by matthewp — APPROVED** [review]

File: packages/integrations/cloudflare/README.md
Comment (alexanderniebuhr) [reply packages/integrations/cloudflare/README.md:53]: @sarah11918 thanks, some technical details are not quite right, so I tried to update that with you suggestions in mind.

Also the build configuration option is not added by me, it already exists in core.

File: packages/integrations/cloudflare/README.md
Comment (sarah11918) [inline packages/integrations/cloudflare/README.md:57]:
```suggestion
With the build configuration `split: true`, the adapter instead compiles a separate bundle for each page. This option requires some manual maintenance of the `functions` folder. Files emitted by Astro will overwrite existing `functions` files with identical names, so you must choose unique file names for each file you manually add. Additionally, the adapter will never empty the `functions` folder of outdated files, so you must clean up the folder manually when you remove pages.

Note that this adapter does not support using [Cloudflare Pages Middleware](https://developers.cloudflare.com/pages/platform/functions/middleware/). Astro will bundle the [Astro middleware](https://docs.astro.build/en/guides/middleware/) into each page.
```
Tiny nits to fix, including I just didn't like having "Additionally," twice in a row like that!  :sweat_smile: (Not wrong, just trying to go for some sentence variety. Either one could be changed!)

Also, the link to writing Cloudflare custom functions now feels a little out of place? Can we get away without it entirely?

Otherwise, happy to sign off on this form of the document!

Comment (alexanderniebuhr) [reply packages/integrations/cloudflare/README.md]: Makes sense :)

I think we can get away without the link, because the Middleware link also links to their docs. People writing custom functions, will probably know the Cloudflare Docs anyways.

**Review by sarah11918 — APPROVED** [review]
Docs is happy! :partying_face:

**Pull request merged**
